core/drivers/clk/clk.c

```diff
@@ -0,0 +1,279 @@
+{
+	return clk->num_parents;
+}
+
+struct clk *clk_get_parent_by_index(struct clk *clk, unsigned int pidx)
+{
+	if (pidx >= clk->num_parents)
+		return NULL;
+
+	return clk->parents[pidx];
+}
+
+static void clk_init_parent(struct clk *clk)
+{
+	uint8_t pidx = 0;
+
+	switch (clk->num_parents) {
+	case 0:
+		break;
+	case 1:
+		clk->parent = clk->parents[0];
+		break;
+	default:
+		pidx = clk->ops->get_parent(clk);
+		if (pidx >= clk->num_parents) {
+			EMSG("get_parent returned an invalid value");
+			panic();
+		}
+
+		clk->parent = clk->parents[pidx];
+		break;
+	}
+}
+
+TEE_Result clk_register(struct clk *clk)
+{
+	if (!clk_check(clk))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	SLIST_INSERT_HEAD(&clk_list, clk, link);
+
+	clk_init_parent(clk);
+	clk_compute_rate(clk);
+
+	DMSG("Registered clock %s, freq %ld", clk->name, clk_get_rate(clk));
+
+	return TEE_SUCCESS;
+}
+
+unsigned long clk_get_rate(struct clk *clk)
+{
+	return clk->rate;
+}
+
+static TEE_Result clk_enable_no_lock(struct clk *clk)
+{
+	struct clk *parent;
+
+	if (refcount_inc(&clk->enabled_count))
+		return TEE_SUCCESS;
+
+	parent = clk_get_parent(clk);
+	if (parent)
+		clk_enable(parent);
+
+	if (clk->ops->enable) {
+		if (clk->ops->enable(clk))
+			return TEE_ERROR_GENERIC;
+	}
+
+	refcount_set(&clk->enabled_count, 1);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result clk_enable(struct clk *clk)
+{
+	TEE_Result ret;
+
+	cpu_spin_lock(&clk->lock);
+	ret = clk_enable_no_lock(clk);
+	cpu_spin_unlock(&clk->lock);
+
+	return ret;
+}
+
+static void clk_disable_no_lock(struct clk *clk)
+{
+	struct clk *parent;
+
+	if (!refcount_dec(&clk->enabled_count))
+		return;
+
+	if (clk->ops->disable)
+		clk->ops->disable(clk);
+
+	parent = clk_get_parent(clk);
+	if (parent)
+		clk_disable(parent);
+}
+
+void clk_disable(struct clk *clk)
+{
+	cpu_spin_lock(&clk->lock);
+	clk_disable_no_lock(clk);
+	cpu_spin_unlock(&clk->lock);
+}
+
+static TEE_Result clk_set_rate_no_lock(struct clk *clk, unsigned long rate)
+{
+	unsigned long parent_rate = 0;
+
+	if (clk->ops->set_rate) {
+		if (clk->parent)
+			parent_rate = clk->parent->rate;
+
+		if (clk->ops->set_rate(clk, rate, parent_rate))
+			return TEE_ERROR_GENERIC;
+	}
+
+	clk_compute_rate(clk);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result clk_set_rate(struct clk *clk, unsigned long rate)
+{
+	TEE_Result ret = TEE_SUCCESS;
+
+	cpu_spin_lock(&clk->lock);
+	if (rate == clk->rate)
+		goto out;
+
+	if (clk->flags & CLK_SET_RATE_GATE) {
+		if (clk_is_enabled(clk)) {
+			ret = TEE_ERROR_BAD_STATE;
+			goto out;
+		}
+	}
+
+	ret = clk_set_rate_no_lock(clk, rate);
+out:
+	cpu_spin_unlock(&clk->lock);
+
+	return ret;
+}
+
+static TEE_Result clk_set_parent_no_lock(struct clk *clk, unsigned int pidx)
+{
+	TEE_Result ret = TEE_SUCCESS;
+	bool was_enabled = false;
+
+	/* Requested parent is already the one set */
+	if (clk->parent == clk->parents[pidx])
+		return TEE_SUCCESS;
+
+	was_enabled = clk_is_enabled(clk);
+	if (was_enabled)
+		clk_disable_no_lock(clk);
```

Comment:
this does not ensure the clock is disabled, there maybe other refcounts accumulated. This should return an error, as done in `clk_set_rate()`

Reply:
This call to disable is meant to maintain the refcount to a correct value not to disable the clock. If the clock needs to be disabled while being modified, the driver should handle that. But CLK_SET_RATE_GATE is there for that kind of scenarios where the clock should not be modified with enable peripheral.

Reply:
ok, i understand. thanks.
I suggest an short inline comment like `/* Decrement refcounts while on old parent clock */`
and `/* Re-increment refcounts with the new parent clock */` at line 253.

Or maybe no inline comment but rename variable `was_enabled` to `update_refcounts`.

Reply:
I guess the comment is more appropriate here, I'll update that.